enhancements/network/ovn-hypershift-live-migration.md

```diff
@@ -0,0 +1,287 @@
+│   dns_server: [kubedns]         │
+│   server_id: 169.254.1.1        │
+│   server_mac: c0:ff:ee:00:00:01 │
+└─────────────────────────────────┘
+```
+
+**Virt-launcher pod address:**
+ The CNI will not set an address at virt-launcher pod netns, that address is
+ assigned to the VM with the DHCP options from the LSP, this allows to use 
+ kubevirt bridge binding with pod networking and still do live migration.
+
+#### IPAM
+
+The /32 routing feature allows an address to be runnig at a node different from
+the one "owning" the subnet the address is comming from. This will happend
+after VM live migration.
+
+One scenario for live migration is to shut down the node the VMs were migrated
+from, this means that the IPAM node subnet should go back to the pool but since
+the migrated VMs contains IPs from it those IPs should reserved in case the 
+subnet is assigned to a new node.
+
+On that case before assigning the subnet to the node the VMs ips need to be 
+reserved so they don't get assigned to new pods
+
+Another scenario is ovn-kubernetes pods restarting after live migration, on 
+that case ovnkube-master should discover to what ip pool the VM belongs 
+(using an annotation or searching for subnets) and reserve the address.
+
+#### Detecting hypershift VMs and live migration
+
+This enhancement only applies for hypershift kubevirt workers so the rest of 
+virtual machine should not use it, to do a new label will be added to
+virt-launcher pod so ovn-k can detect if the /32 routing need to be done.
+
+There is also a moment where routing need to do the "switch", meaning that the
+vm has to be routed over the host is running on, to detect that the label
+"kubevirt.io/nodeName" can be check since it appears just after live migration
+has finish. Also to discard old virt-launcher pod the creation timestamp can
+be used, ignoring them is they are not the latest one.
+
+### Risks and Mitigations
+
+### OVN extended arp_proxy features ####
+
+
+The current ovn arp_proxy is missing the following features: enforcing the mac 
+address and ipv6, ideally those get implemented and ovn-kubernetes have 
+them for 4.14 release.
+
+In case they don't arrive on time, it will not be able to configure IPv6 hosted
+clusters at hypershift and VMs arp table will need to be refreshed after live
+migration. 
+
+### Open Questions 
+
+Do this impact non hypershift payloads ? The control-plane will have some more
+load from the /32 routing but data-plane is not affected
+
+Do this work with future interconnect topology ? It's not using a distributed 
+switch per hosted closter so that part should be fine, but it may needed 
+to replicate /32 routing at all node's ovn_cluster_router.
+
+Do this need extra configuration at openshift virtualization ? The bridge 
+binding pod network live migration feature gate need to be activated
+
+### Test Plan
+
+Conformance for the hosted cluster has to be run before, during and after live
+migration.
```

Comment:
I think we'll need to add a conformance test that explicitly exercises that connections are not dropped during live migration.

We actually have the ability to add this kind of test to our conformance test suite `openshift/origin`. With knowledge that we're running the `kubevirt` infrastructure, and with access to the underlying infra cluster's kubeconfig, we can force a live migration to occur while holding tcp connections open between two pods on separate kubevirt vm nodes.

see Oren's PR here [1] for more info. Once that PR is merged, all the pieces will exist for a live migration test. We'll have the kubeconfigs for infra and guest clusters, and be able to execute the live migration from the test driver

1. https://github.com/openshift/origin/pull/27456